File: frontend/src/lib/api/studio/websites.ts
/**
 * Websites API
 * Educational Note: Handles AI-generated multi-page websites.
 * Uses an agentic approach for complex multi-file generation.
 */

import axios from 'axios';
import { API_BASE_URL } from '../client';
import type { JobStatus } from './index';
import { createLogger } from '@/lib/logger';

const log = createLogger('studio-websites-api');

/**
 * Website page information
 */
export interface WebsitePage {
  filename: string;
  page_title: string;
  description: string;
}

/**
 * Website design system
 */
export interface WebsiteDesignSystem {
  primary_color: string;
  secondary_color: string;
  accent_color?: string;
  background_color: string;
  text_color: string;
  font_family: string;
}

/**
 * Generated website image
 */
export interface WebsiteImage {
  purpose: string;
  filename: string;
  placeholder: string;
  url: string;
}

/**
 * Website generation job
 */
export interface WebsiteJob {
  id: string;
  source_id: string;
  source_name: string;
  direction: string;
  status: JobStatus;
  status_message: string;
  error_message: string | null;
  // Edit lineage
  parent_job_id: string | null;
  edit_instructions: string | null;

  // Plan
  site_type: 'portfolio' | 'business' | 'blog' | 'landing' | 'corporate' | 'personal' | 'ecommerce' | null;
  site_name: string | null;
  pages: WebsitePage[];
  features: string[];
  design_system: WebsiteDesignSystem | null;
  navigation_style: 'fixed' | 'sticky' | 'static' | null;
  images_needed: Array<{
    purpose: string;
    description: string;
    aspect_ratio: string;
  }>;
  layout_notes: string | null;

  // Generated content
  images: WebsiteImage[];
  files: string[];
  pages_created: string[];
  features_implemented: string[];
  cdn_libraries_used: string[];
  summary: string | null;

  // URLs
  preview_url: string | null;
  download_url: string | null;

  // Metadata
  iterations: number | null;
  input_tokens: number | null;
  output_tokens: number | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/**
 * Response from starting website generation
 */
export interface StartWebsiteResponse {
  success: boolean;
  job_id?: string;
  status?: string;
  message?: string;
  error?: string;
}

/**
 * Response from getting website job status
 */
export interface WebsiteJobStatusResponse {
  success: boolean;
  job?: WebsiteJob;
  error?: string;
}

/**
 * Response from listing website jobs
 */
export interface ListWebsiteJobsResponse {
  success: boolean;
  jobs: WebsiteJob[];
  error?: string;
}

/**
 * Websites API
 */
export const websitesAPI = {
  /**
   * Start website generation or edit (background task)
   * Educational Note: Non-blocking - returns immediately with job_id
   */
  async startGeneration(
    projectId: string,
    sourceId: string,
    direction?: string,
    parentJobId?: string,
    editInstructions?: string
  ): Promise<StartWebsiteResponse> {
    try {
      const body: Record<string, unknown> = {
        source_id: sourceId,
        direction: direction || '',
      };
      if (parentJobId) body.parent_job_id = parentJobId;
      if (editInstructions) body.edit_instructions = editInstructions;

      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/studio/website`,
        body
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to start website generation');
      throw error;
    }
  },

  /**
   * Get website job status
   */
  async getJobStatus(projectId: string, jobId: string): Promise<WebsiteJobStatusResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/website-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to get website job status');
      throw error;
    }
  },

  /**
   * List all website jobs for a project, optionally filtered by source
   */
  async listJobs(projectId: string, sourceId?: string): Promise<ListWebsiteJobsResponse> {
    try {
      const params = sourceId ? { source_id: sourceId } : {};
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/website-jobs`,
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to list website jobs');
      throw error;
    }
  },

  /**
   * Get the preview URL for a website (opens in new window)
   */
  getPreviewUrl(projectId: string, jobId: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/websites/${jobId}/preview`;
  },

  /**
   * Get the download URL for a website (ZIP with all files)
   */
  getDownloadUrl(projectId: string, jobId: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/websites/${jobId}/download`;
  },

  /**
   * Delete a website job
   */
  async deleteJob(projectId: string, jobId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/studio/website-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to delete website job');
      throw error;
    }
  },

  /**
   * Poll website job status until complete or error
   */
  async pollJobStatus(
    projectId: string,
    jobId: string,
    onProgress?: (job: WebsiteJob) => void,
    intervalMs: number = 2000,
    maxAttempts: number = 200  // Websites can take longer (multi-file, iterative)
  ): Promise<WebsiteJob> {
    let attempts = 0;
    let currentInterval = intervalMs;

    while (attempts < maxAttempts) {
      const response = await this.getJobStatus(projectId, jobId);

      if (!response.success || !response.job) {
        throw new Error(response.error || 'Failed to get job status');
      }

      const job = response.job;

      if (onProgress) {
        onProgress(job);
      }

      if (job.status === 'ready' || job.status === 'error') {
        return job;
      }

      await new Promise((resolve) => setTimeout(resolve, currentInterval));

      attempts++;

      if (attempts > 5 && currentInterval < 5000) {
        currentInterval = Math.min(currentInterval * 1.2, 5000);
      }
    }

    throw new Error('Website generation timed out');
  },
};

File: backend/app/api/studio/websites.py
"""
Website Generator endpoints - AI-generated multi-page websites.

Educational Note: Website generation demonstrates complex agent workflows:
1. website_agent_executor orchestrates the entire process
2. Claude generates HTML, CSS, and JavaScript
3. Gemini generates images for the site
4. Complete package: multiple pages + assets

Agent Architecture:
- Uses website_agent_executor for orchestration
- Agent has tools for page creation, styling, scripting
- Multi-page sites with navigation
- Responsive design patterns

Output Structure:
- index.html (main page)
- Additional pages (about.html, etc.)
- styles.css (stylesheet)
- script.js (JavaScript)
- assets/ folder for images
- All served via Flask endpoints

Routes:
- POST /projects/<id>/studio/website                        - Start generation
- GET  /projects/<id>/studio/website-jobs/<id>              - Job status
- GET  /projects/<id>/studio/website-jobs                   - List jobs
- GET  /projects/<id>/studio/websites/<id>/<path:file>      - Serve files
- GET  /projects/<id>/studio/websites/<id>/preview          - Preview site
- GET  /projects/<id>/studio/websites/<id>/download         - Download ZIP
"""
import io
import re
import zipfile
from flask import jsonify, request, current_app, send_file, Response
from app.api.studio import studio_bp
from app.services.studio_services import studio_index_service
from app.services.integrations.supabase import storage_service
from app.services.auth import require_permission


@studio_bp.route('/projects/<project_id>/studio/website', methods=['POST'])
@require_permission("studio", "websites")
def generate_website(project_id: str):
    """
    Start website generation or edit (background task).

    Request body:
        {
            "source_id": "source-uuid",
            "direction": "optional user direction/preferences",
            "parent_job_id": "optional parent job UUID for edits",
            "edit_instructions": "optional edit instructions"
        }

    Returns:
        202 Accepted with job_id for polling
    """
    from app.services.tool_executors import website_agent_executor

    try:
        data = request.get_json()
        source_id = data.get('source_id')
        direction = data.get('direction', '')

        # Edit mode: load parent job's files as context for refinement
        parent_job_id = data.get('parent_job_id')
        edit_instructions = data.get('edit_instructions')
        previous_markdown = None
        previous_title = None
        parent_source_name = None

        if parent_job_id:
            # Clean up any previously failed edit jobs for this parent
            try:
                all_jobs = studio_index_service.list_website_jobs(project_id)
                for job in all_jobs:
                    if (job.get("status") == "error"
                            and job.get("parent_job_id") == parent_job_id):
                        studio_index_service.delete_website_job(project_id, job["id"])
            except Exception:
                pass  # Non-critical cleanup — don't block the new edit

            parent_job = studio_index_service.get_website_job(project_id, parent_job_id)
            if not parent_job or not parent_job.get('files'):
                return jsonify({
                    'success': False,
                    'error': 'Parent job not found or has no content to edit'
                }), 404

            # Collect all file content from Supabase Storage
            file_contents = []
            for fname in (parent_job.get('files') or []):
                content = storage_service.download_studio_file(
                    project_id=project_id,
                    job_type="websites",
                    job_id=parent_job_id,
                    filename=fname
                )
                if content:
                    file_contents.append(f"--- {fname} ---\n{content}")
            previous_markdown = "\n\n".join(file_contents) if file_contents else None
            if previous_markdown is None:
                return jsonify({
                    'success': False,
                    'error': 'Failed to load parent website content from storage'
                }), 500
            previous_title = parent_job.get('site_name')
            parent_source_name = parent_job.get('source_name')
        elif not source_id:
            return jsonify({
                'success': False,
                'error': 'source_id is required'
            }), 400

        # Execute website generation (background task)
        result = website_agent_executor.execute(
            project_id=project_id,
            source_id=source_id or '',
            direction=direction,
            edit_instructions=edit_instructions,
            previous_markdown=previous_markdown,
            previous_title=previous_title,
            parent_job_id=parent_job_id,
            parent_source_name=parent_source_name
        )

        if not result.get('success'):
            return jsonify(result), 400

        return jsonify(result), 202  # Accepted

    except Exception as e:
        current_app.logger.error(f"Error starting website generation: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to start website generation: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/website-jobs/<job_id>', methods=['GET'])
def get_website_job_status(project_id: str, job_id: str):
    """
    Get status of a website generation job.

    Returns:
        Job object with current status, progress, and results if complete
    """
    try:
        job = studio_index_service.get_website_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': 'Job not found'
            }), 404

        return jsonify({
            'success': True,
            'job': job
        })

    except Exception as e:
        current_app.logger.error(f"Error getting website job status: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to get job status: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/website-jobs', methods=['GET'])
def list_website_jobs(project_id: str):
    """
    List all website jobs for a project, optionally filtered by source.

    Query params:
        source_id (optional): Filter by source ID

    Returns:
        List of website jobs sorted by created_at descending
    """
    try:
        source_id = request.args.get('source_id')
        jobs = studio_index_service.list_website_jobs(project_id, source_id)

        # Filter out orphaned failed-edit jobs (error + parent_job_id).
        clean_jobs = [
            job for job in jobs
            if not (job.get("status") == "error" and job.get("parent_job_id"))
        ]

        return jsonify({
            'success': True,
            'jobs': clean_jobs
        })

    except Exception as e:
        current_app.logger.error(f"Error listing website jobs: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to list jobs: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/websites/<job_id>/<path:filename>', methods=['GET'])
def get_website_file(project_id: str, job_id: str, filename: str):
    """
    Serve a website file (HTML, CSS, JS, or image asset).

    Supports:
        - HTML pages: index.html, about.html, etc.
        - Stylesheets: styles.css
        - Scripts: script.js
        - Assets: assets/image_1.png, etc.
    """
    try:
        # Determine MIME type
        mime_type = 'text/html'
        if filename.endswith('.css'):
            mime_type = 'text/css'
        elif filename.endswith('.js'):
            mime_type = 'application/javascript'
        elif filename.endswith('.png'):
            mime_type = 'image/png'
        elif filename.endswith('.jpg') or filename.endswith('.jpeg'):
            mime_type = 'image/jpeg'
        elif filename.endswith('.gif'):
            mime_type = 'image/gif'
        elif filename.endswith('.svg'):
            mime_type = 'image/svg+xml'
        elif filename.endswith('.webp'):
            mime_type = 'image/webp'

        # Binary files (images)
        is_binary = mime_type.startswith('image/')
        if is_binary:
            file_data = storage_service.download_studio_binary(
                project_id, "websites", job_id, filename
            )
            if file_data is None:
                return jsonify({'success': False, 'error': 'File not found'}), 404
            return send_file(io.BytesIO(file_data), mimetype=mime_type, as_attachment=False)

        # Text files (HTML, CSS, JS)
        content = storage_service.download_studio_file(
            project_id, "websites", job_id, filename
        )
        if content is None:
            return jsonify({'success': False, 'error': 'File not found'}), 404

        # For CSS files, inject auth token into url() references so images load.
        # Educational Note: CSS url() references (e.g. background-image) trigger
        # separate browser requests that don't carry the parent page's query params.
        token = request.args.get('token', '')
        if mime_type == 'text/css' and token:
            def _add_token_to_css_url(match: re.Match) -> str:
                prefix = match.group(1)
                url = match.group(2)
                suffix = match.group(3)
                sep = '&' if '?' in url else '?'
                return f'{prefix}{url}{sep}token={token}{suffix}'

            content = re.sub(
                r"""(url\(["']?)(?!https?://|//|data:)([^"')\s]+)(["']?\))""",
                _add_token_to_css_url,
                content
            )

        return Response(content, mimetype=mime_type)

    except Exception as e:
        current_app.logger.error(f"Error serving website file: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to serve file: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/websites/<job_id>/preview', methods=['GET'])
def preview_website(project_id: str, job_id: str):
    """
    Preview website by serving index.html with auth tokens injected.

    Educational Note: The iframe can't pass Authorization headers for sub-resource
    requests (CSS, JS, images). We inject ?token= into local resource URLs in the
    HTML so the browser passes the JWT when fetching these files. The original
    files on disk stay clean for download/export.
    """
    try:
        # Download index.html from Supabase Storage
        html_content = storage_service.download_studio_file(
            project_id, "websites", job_id, "index.html"
        )
        if html_content is None:
            return jsonify({
                'success': False,
                'error': 'Website not ready yet or index.html not found'
            }), 404

        # Inject auth token into local resource URLs (CSS, JS, images)
        token = request.args.get('token', '')
        if token:
            def _add_token(match: re.Match) -> str:
                attr = match.group(1)   # src or href
                url = match.group(2)
                sep = '&' if '?' in url else '?'
                return f'{attr}="{url}{sep}token={token}"'

            # Match src="..." and href="..." but skip external URLs
            html_content = re.sub(
                r'(src|href)="(?!https?://|//|data:|#|mailto:)([^"]+)"',
                _add_token,
                html_content
            )

        return Response(html_content, mimetype='text/html')

    except Exception as e:
        current_app.logger.error(f"Error previewing website: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to preview website: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/websites/<job_id>/download', methods=['GET'])
def download_website(project_id: str, job_id: str):
    """
    Download website as ZIP file containing all files.

    Returns:
        ZIP file with all HTML pages, CSS, JS, and image assets
    """
    try:
        # Get job info
        job = studio_index_service.get_website_job(project_id, job_id)
        if not job:
            return jsonify({
                'success': False,
                'error': 'Job not found'
            }), 404

        if job['status'] != 'ready':
            return jsonify({
                'success': False,
                'error': 'Website not ready yet'
            }), 400

        site_name = job.get('site_name', 'Website')
        zip_filename = f"{site_name.replace(' ', '_')}.zip"

        # Create ZIP in memory from Supabase Storage
        zip_buffer = io.BytesIO()

        with zipfile.ZipFile(zip_buffer, 'w', zipfile.ZIP_DEFLATED) as zip_file:
            # Add text files (HTML, CSS, JS) from job metadata
            for fname in job.get('files') or []:
                content = storage_service.download_studio_file(
                    project_id, "websites", job_id, fname
                )
                if content:
                    zip_file.writestr(fname, content)

            # Add image assets
            for image_info in job.get('images') or []:
                img_filename = image_info.get('filename')
                if img_filename:
                    img_data = storage_service.download_studio_binary(
                        project_id, "websites", job_id, f"assets/{img_filename}"
                    )
                    if img_data:
                        zip_file.writestr(f"assets/{img_filename}", img_data)

        zip_buffer.seek(0)

        return send_file(
            zip_buffer,
            mimetype='application/zip',
            as_attachment=True,
            download_name=zip_filename
        )

    except Exception as e:
        current_app.logger.error(f"Error downloading website: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to download website: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/website-jobs/<job_id>', methods=['DELETE'])
def delete_website_job(project_id: str, job_id: str):
    """
    Delete a website job and its files from Supabase Storage.

    Response:
        - Success status
    """
    try:
        job = studio_index_service.get_website_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Website job {job_id} not found'
            }), 404

        # Delete all files for this job from Supabase Storage
        storage_service.delete_studio_job_files(
            project_id=project_id,
            job_type="websites",
            job_id=job_id
        )

        # Delete job from index
        studio_index_service.delete_website_job(project_id, job_id)

        return jsonify({
            'success': True,
            'message': f'Website job {job_id} deleted'
        })

    except Exception as e:
        current_app.logger.error(f"Error deleting website job: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to delete website job: {str(e)}'
        }), 500
